File: java/Player.java
import java.util.Random;

public abstract class Player {
	private int MOBILITY;
	protected Pos pos;
    protected int healthCap;
	protected int health;
	protected Object equipment;
    protected Wand wand;
	protected int index;
	protected String myString;
	protected SurvivalGame game;
	
	public Player(int healthCap, int mob, int posx, int posy, int index, SurvivalGame game) {

		this.MOBILITY = mob;
        this.healthCap = healthCap;
		this.health = this.healthCap;
		this.pos = new Pos(posx, posy);
		this.index = index;
		this.game = game;
	}
    
    public int getHealth(){
        return this.health;
    }
    
    public String getString(){
        return this.myString;
    }
    public int getIndex(){
        return this.index;
    }
    
    public void setString(String str){
        this.myString=str;
    }

	public Pos getPos() {
		return pos;
	}

	public void teleport() {

		Random rand;
		rand = new Random();
		int randx = rand.nextInt(game.D);
		int randy = rand.nextInt(game.D);
		while (game.positionOccupied(randx, randy)) {
			randx = rand.nextInt(game.D);
			randy = rand.nextInt(game.D);
		}
		pos.setPos(randx, randy);
	}

	public void increaseHealth(int h) {
        if((this.health+h)>this.healthCap){
            this.health=this.healthCap;
        }
        else
            this.health += h;
	}

	public void decreaseHealth(int h) {
		this.health -= h;
		if (this.health <= 0){
            this.health=0;
			this.myString = "C" + this.myString.charAt(0);
        }
	}

	public String getName() {
		return myString;
	}

	public void askForMove() {
		// Print general information
		System.out.println("Your health is " + health
				+ String.format(". Your position is (%d,%d). Your mobility is %d.", pos.getX(), pos.getY(), this.MOBILITY));

		System.out.println("You now have following options: ");
		System.out.println("1. Move");
		System.out.println("2. Attack");
		System.out.println("3. End tne turn");

		int a = SurvivalGame.reader.nextInt();

		if (a == 1) {
			System.out.println("Specify your target position (Input 'x y').");
			int posx = SurvivalGame.reader.nextInt(), posy = SurvivalGame.reader.nextInt();
			if (pos.distance(posx, posy) > this.MOBILITY) {
				System.out.println("Beyond your reach. Staying still.");
			} else if (game.positionOccupied(posx, posy)) {
				System.out.println("Position occupied. Cannot move there.");
			} else {
				this.pos.setPos(posx, posy);
				game.printBoard();
				System.out.println("You can now \n1.attack\n2.End the turn");
				if (SurvivalGame.reader.nextInt() == 1) {
					System.out.println("Input position to attack. (Input 'x y')");
					int attx = SurvivalGame.reader.nextInt(), atty = SurvivalGame.reader.nextInt();
                    Weapon weapon = (Weapon)this.equipment;
					weapon.action(attx, atty);
				}
			}
		} else if (a == 2) {
			System.out.println("Input position to attack.");
			int attx = SurvivalGame.reader.nextInt(), atty = SurvivalGame.reader.nextInt();
            Weapon weapon = (Weapon)this.equipment;
			weapon.action(attx, atty);
		} else if (a == 3) {
			return;
		}
	}
    
    public void askForMoveHeal(){
        System.out.println("Your health is " + health
				+ String.format(". Your position is (%d,%d). Your mobility is %d.", pos.getX(), pos.getY(), this.MOBILITY));

		System.out.println("You now have following options: ");
		System.out.println("1. Move");
		System.out.println("2. Heal");
		System.out.println("3. End tne turn");

		int a = SurvivalGame.reader.nextInt();
        
        if (a == 1) {
			System.out.println("Specify your target position (Input 'x y').");
			int posx = SurvivalGame.reader.nextInt(), posy = SurvivalGame.reader.nextInt();
			if (pos.distance(posx, posy) > this.MOBILITY) {
				System.out.println("Beyond your reach. Staying still.");
			} else if (game.positionOccupied(posx, posy)) {
				System.out.println("Position occupied. Cannot move there.");
			} else {
				this.pos.setPos(posx, posy);
				game.printBoard();
				System.out.println("You can now \n1.heal\n2.End the turn");
				if (SurvivalGame.reader.nextInt() == 1) {
					System.out.println("Input position to heal. (Input 'x y')");
					int healx = SurvivalGame.reader.nextInt(), healy = SurvivalGame.reader.nextInt();
					this.wand.action(healx, healy);
				}
			}
		} else if (a == 2) {
			System.out.println("Input position to heal.");
			int healx = SurvivalGame.reader.nextInt(), healy = SurvivalGame.reader.nextInt();
			this.wand.action(healx, healy);
		} else if (a == 3) {
			return;
		}
    }

}

File: python/task4/Player.py
import random
from Pos import Pos
class Player(object):
    def __init__(self,healthCap,mob,posx,posy,index,game):
        self.MOBILITY=mob
        self.healthCap=healthCap
        self.health=self.healthCap
        self.pos=Pos(posx,posy)
        self.equipment=0
        self.index=index
        self.myString=""
        self.game=game
        
    def getPos(self):
        return self.pos
    
    def teleport(self):
        randx=random.randint(0,(self.game.D-1))
        randy=random.randint(0,(self.game.D-1))
        while(self.game.positionOccupied(randx,randy)):
            randx=random.randint(0,(self.game.D-1))
            randy=random.randint(0,(self.game.D-1))
        self.pos.setPos(randx,randy)
    
    def increaseHealth(self,h):
        if((self.health+h)>self.healthCap):
            self.health=self.healthCap
        else:
            self.health+=h
    
    def decreaseHealth(self,h):
        self.health-=h
        if (self.health<=0):
            self.health=0
            self.myString = "C" + self.myString[0]
            
    def getName(self):
        return self.myString
    
    def askForMove(self):
        string="Your health is "+str(self.health)+". Your position is ("+str(self.pos.getX())+","+str(self.pos.getY())+"). Your mobility is "+str(self.MOBILITY)+"."
        print string
        print "You now have following options: "
        print "1. Move"
        print "2. Attack"
        print "3. End the turn"
        a=input()
        
        if(a==1):
            print "Specify your target position (Input 'x y')."
            x, y = raw_input().split()
            posx=int(x)
            posy=int(y)
            if(self.pos.distance(posx,posy)>self.MOBILITY):
                print "Beyond your reach. Staying still."
            elif(self.game.positionOccupied(posx,posy)):
                print "Position occupied. Cannot move there."
            else:
                self.pos.setPos(posx,posy)
                self.game.printBoard()
                print "You can now \n1.attack\n2.End the turn"
                choice=input()
                if(choice==1):
                    print "Input position to attack. (Input 'x y')"
                    ax, ay = raw_input().split()
                    attx=int(ax)
                    atty=int(ay)
                    self.equipment.action(attx,atty)
        elif(a==2):
            print "Input position to attack."
            ax, ay = raw_input().split()
            attx=int(ax)
            atty=int(ay)
            self.equipment.action(attx,atty)
        elif(a==3):
            return
    
    def askForMoveHeal(self):
        string="Your health is "+str(self.health)+". Your position is ("+str(self.pos.getX())+","+str(self.pos.getY())+"). Your mobility is "+str(self.MOBILITY)+"."
        print string
        print "You now have following options: "
        print "1. Move"
        print "2. Heal"
        print "3. End the turn"
        a=input()
        
        if(a==1):
            print "Specify your target position (Input 'x y')."
            x, y = raw_input().split()
            posx=int(x)
            posy=int(y)
            if(self.pos.distance(posx,posy)>self.MOBILITY):
                print "Beyond your reach. Staying still."
            elif(self.game.positionOccupied(posx,posy)):
                print "Position occupied. Cannot move there."
            else:
                self.pos.setPos(posx,posy)
                self.game.printBoard()
                print "You can now \n1.heal\n2.End the turn"
                choice=input()
                if(choice==1):
                    print "Input position to heal. (Input 'x y')"
                    hx, hy = raw_input().split()
                    healx=int(hx)
                    healy=int(hy)
                    self.equipment.action(healx,healy)
        elif(a==2):
            print "Input position to heal."
            hx, hy = raw_input().split()
            healx=int(hx)
            healy=int(hy)
            self.equipment.action(healx,healy)
        elif(a==3):
            return